network/topics/scoring.go

```diff
@@ -26,52 +25,128 @@ func DefaultScoringConfig() *ScoringConfig {
 
 // scoreInspector inspects scores and updates the score index accordingly
 // TODO: finalize once validation is in place
-func scoreInspector(logger *zap.Logger, scoreIdx peers.ScoreIndex, logFrequency int, metrics Metrics, peerConnected func(pid peer.ID) bool) pubsub.ExtendedPeerScoreInspectFn {
+func scoreInspector(logger *zap.Logger, scoreIdx peers.ScoreIndex, logFrequency int, metrics Metrics, peerConnected func(pid peer.ID) bool, peerScoreParams *pubsub.PeerScoreParams, topicScoreParamsFactory func(string) *pubsub.TopicScoreParams) pubsub.ExtendedPeerScoreInspectFn {
 	inspections := 0
 
 	return func(scores map[peer.ID]*pubsub.PeerScoreSnapshot) {
+
+		if inspections%logFrequency != 0 {
+			// Don't log yet.
+			inspections++
+			return
+		}
+
 		// Reset metrics before updating them.
 		metrics.ResetPeerScores()
 
+		// Use a "scope cache" for getting a topic's score parameters
+		// otherwise, the factory method would be called multiple times for the same topic
+		topicScoreParamsCache := make(map[string]*pubsub.TopicScoreParams)
+		getScoreParamsForTopic := func(topic string) *pubsub.TopicScoreParams {
+			if topicParams, exists := topicScoreParamsCache[topic]; exists {
+				return topicParams
+			}
+			topicScoreParamsCache[topic] = topicScoreParamsFactory(topic)
+			return topicScoreParamsCache[topic]
+		}
```

Comment:
can score params for a certain topic change? if so the cache will not reflect that. I know for fact they change in fork so we can include special code for that like in #1477 , if there aren't any other cases then we can resolve this.

Reply:
The p2 parameters may change upon validators' addition/removal.

This cache has only the scope of one execution of the function. So, for the next execution, there will be a new cache and it will use the updated parameters.

I added this cache since we would generate the same topic parameters repeatedly for each peer and waste resource (I mean that in a single function execution).

Does the code seem aligned with my purpose?